chain/chain/src/chain.rs

```diff
@@ -1524,6 +1524,9 @@ impl Chain {
         let state_part =
             self.runtime_adapter.obtain_state_part(shard_id, &state_root, part_id, num_parts);
 
+        // Before saving State Part data, we need to make sure we can calculate and save State Header
+        self.get_state_response_header(shard_id, sync_hash)?;
```

Comment:
Can we just check whether key exists in `ColStateHeader`? If you call `get_state_response_header` it will calculate the header in case header doesn't exist.

Reply:
We need to have the Header to know the number of Parts.
In average, we need to calculate Header anyway to response for State Requests.